### PR TITLE
Add pieces level for representing various game pieces 

### PR DESCRIPTION
**Pull request by MaskyYoung**
Hi Amanpreet,

We wrote the logic for pieces adding. But there are something wrong in the drop function so that when we drop the second item on the first item on the board, there will create a second piece type in the piece table. We tried our best to find out the faults in the logic but we failed. Do you mind helping us reviewing our code and try to figure out how this happened?
Thank you!

### CODE REVIEW

**Comment by apsdehal** [comment]
Can you write the steps to reproduce the bug?

**Comment by MaskyYoung** [comment]
1. drag a piece of image on board
2. drag the same image on the first one
3.check the pieces on Firebase, the image keys stored in each piece type
are abnormal

Amanpreet Singh <notifications@github.com>于2017年10月1日 周日上午12:53写道：

**Comment by apsdehal** [comment]
Can you ping me at hangouts on my personal email id?

File: src/components/GameSpecBuilder.js
Comment (apsdehal) [inline src/components/GameSpecBuilder.js]: Instead of using both items and pieces, merge them to one, so that it is easier to maintain state.

File: src/components/Login.js
Comment (apsdehal) [inline src/components/Login.js]: Good catch.

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: This is probably wrong. A person can drag an image multiple times on board, so you don't need to compare it like this.

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Keep single state variable, namely pieces which will look something like this:
```js
pieces = {
   images: [],
   offsets: [],
   key: ''
} 
```
Now you will fill them accordingly. So, in a piece image at ith index i.e. `piece.images[i]` will have offset at `piece.offsets[i]`.

The above shown information can be directly stored into firebase. But in an ideal case scenario you should only save images field in pieces for now.

After saving pieces in firebase will look like:

```
pieces 
      key_here
               images: [ image_key_here, image_key2_here ]
```

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: So this forEach will become double forEach.
```js
pieces.forEach((piece, index) => {
     piece.images.forEach((image, index) => {
     });
});
```

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: The value set should be passed as an object so pass `{images}`. Current solution will just save array inside the child with key childKey which is not what we want, we want images inside images field in the piece child. So use `{images}` instead of directly passing images

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Same as above here

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Instead of keeping both keep one with track of images and offsets as said above

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Also check if the piece was already on the board i.e. was present in some piece. Easy way to do this would be to `Piece.js` file in which you should item to the piece in which the image is present and the index of the image in that piece. Also if the piece gets empty delete it from firebase.

Comment (apsdehal) [reply src/components/gamespec/Board.js]: I think you are mostly getting error due to that. You are also not removing the image from piece in which it was already present. So if it was already present in a piece, you should have two firebase sets in total

Comment (apsdehal) [inline src/components/gamespec/Board.js]: Also take care of the space ~`if{`~  -> `if {`

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Write just `x * x`

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing ~`=>{`~ -> `=> {`

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing see above

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Missing indent here

**Review by apsdehal — CHANGES_REQUESTED** [review]
Also provide a better commit message of what this commit actually does. For linting issues, once you make this change I will add eslint to project which will do automatically linting for you guys. Fix the issues, I have asked which are probably causing other issues for you then we can go ahead and merge this one. Also don't merge the pull request on your own. I will do that once I review and approve it. Cheers!

File: src/components/Login.js
Comment (MaskyYoung) [reply src/components/Login.js]: Thank you!

File: src/components/GameSpecBuilder.js
Comment (MaskyYoung) [reply src/components/GameSpecBuilder.js]: Sure, thanks

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: So you mean the pieces should belong to spec instead of being the same level with users, images and specs?

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: can you explain ident here more precisely?

Comment (apsdehal) [reply src/components/gamespec/Board.js]: I fixed the typo

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: Pieces is still a high level object in firebase, I have fixed the above description.

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: Sure, thanks

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: Is there any possibility that one image can belong to many pieces?

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: We are confused that if we only have pieces and delete items, how should we deal with drag action since the index of each item is not absolute any more?

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: one more question: should the image keys present in high level field--pieces be distinct or the keys can be duplicated?

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: Yes, that is an obvious possibility. Consider the game of Go, you can drag same black piece to board multiple times creating new pieces each time. So, in all of these new pieces created black piece is the same image that we dragged again and again on the board.

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: You do multiple for loops as I explained above. Also, you can implement in any way you want, I am just giving possible suggestions. If you want to keep multiple pieces you can go with it.

Furthermore, for unique index you can think of piece and images as a linear array.
Let's say you have two pieces, one with 3 images and other with 2 images.
So the index of each image becomes uniquely as 0, 1, 2, 3, 4 where you can actually flatten the array. There are multiple ways to do this, this is just one of them. You are however, free to implement the functionality however you want as long the final working functionality is the one we expect.

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: keys will the image keys present in the high level firebase field images. So, yes these image keys can come multiple times in single piece's images field.

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: Do you mind explaining again how our error generates cause the first paragraph is hard to understand?

Comment (apsdehal) [reply src/components/gamespec/Board.js]: Were you able to figure out this?

Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: We've solved the problems and pushed the executable code. Could you check the latest commit and examine whether it can be merged?

**Comment by apsdehal** [comment]
I dragged a rook onto the board and then dragged another rook onto the rook already on to board. And finally dragged, king onto the board. I expected that there would be two pieces but there were 3 of them. So, look into the bug. Once this is fixed, I will check other functionalities that are expected. 
See
image
image

**Comment by MaskyYoung** [comment]
You can check the piece type keys, the first piece key equals to the second one. They are still two pieces but 3 images.

**Comment by apsdehal** [comment]
Oh, but ideally you have to only generate two pieces in the final spec. 
The person who is rendering the board needs to check if there are multiple images for a single piece, unless you otherwise specify the imageKey specifically there is no way to map offsets to an image in piece's image list. 
If you do it this way, you need to provide more information on which section the image belongs to. 
But ideally, portal should only get pieceKey and check for images in images field of the piece and get their corresponding offsets from the piece itself.

**Comment by apsdehal** [comment]
Also, I found a small bug in GameBuilder for which I will push one commit to master. Be sure to rebase your branch against it and solve conflicts with master.

**Comment by yangyangluo** [comment]
Sure, we will change the code.

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Fix spacing here

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing in the end here

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Here spacing in multiple places

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Wrong indentation and spacing here also

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Spacing

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Remove all the console logs in final commit

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: A lot of logic in this drag function and the function below is getting reused, extract it out in a single function

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Can you use better variable names, the code is very difficult to read at the moment to understand what is going on

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: This should be array of arrays not an array of objects. The extra layer of object is unnecessary.

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Why are we pushing both 0 and `piece.length - 1` ?

File: src/components/gamespec/Board.js
Comment (apsdehal) [inline src/components/gamespec/Board.js]: Instead of setting a lot of things in multiple if else, set the variable named index and in the end after all of if else set the copy's index variable's information finally. DRY.

**Review by apsdehal — CHANGES_REQUESTED** [review]

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: but I need to return two values if I extract the function which is not a good idea for the javascript

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: Sure, I will try to rename them. Maybe some annotation is useful here?

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: We think the objects can help explain the code well

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: can you explain it clearer?

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: sure

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: I mean you are doing same things like `copy[index].offset.x = position.x` in all of the if else. You can only set it once.

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: I think it got more confusing for me. Took me a while to understand what exactly you are doing.

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: emm, cool

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: Not exactly, you can still return an object.

Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: I think I will need to return the copy and pieces within one function

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: Also, delete the inner arrays which become empty

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: piece.length - 1 means the index of the piece the item belongs to, the 0 means the image index of the item within the piece it belongs to

File: src/components/gamespec/Board.js
Comment (apsdehal) [reply src/components/gamespec/Board.js]: Upto you, I haven't read the code thoroughly, but I can see same (a little bit different) logic in both of them. So they can be shifted to one function.

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: we processed the empty array in the specviewer, if we keep deleting, this will cause a dynamic array and there will be more time complexity to process the drag action. We think leaving the empty array can save a lot of time complexity.

File: src/components/gamespec/Board.js
Comment (MaskyYoung) [reply src/components/gamespec/Board.js]: Yep, we found this out, but we also noticed that if we want to write them into a new function, we need to return two values, which is not recommended on google

**Comment by apsdehal** [comment]
I am going to test this one thoroughly today, make sure you do that too and then I will merge it into master. Then, you guys can start working on script.

**Comment by MaskyYoung** [comment]
Sure, thank you!

Amanpreet Singh <notifications@github.com>于2017年10月10日 周二下午12:12写道：

**Pull request merged**
